constants/constants.go

```diff
@@ -51,14 +27,46 @@ var (
 	// MagicDirectives are directives automatically appended to Dockerfiles
 	// when pushing the image. These directives allow the built image to be
 	// 're-used'.
-	MagicDirectives = fmt.Sprintf(`
-COPY --chmod=0755 %[1]s %[2]s
-COPY --chmod=0644 %[3]s %[4]s
+	MagicDirectives = `
+COPY --chmod=0755 .envbuilder.tmp/envbuilder /.envbuilder/bin/envbuilder
+COPY --chmod=0644 .envbuilder.tmp/image /.envbuilder/image
 USER root
 WORKDIR /
-ENTRYPOINT [%[2]q]
-`,
-		".envbuilder.tmp/envbuilder", MagicBinaryLocation,
-		".envbuilder.tmp/image", MagicImage,
-	)
+ENTRYPOINT ["/.envbuilder/bin/envbuilder"]
+`
 )
+
+// ErrNoFallbackImage is returned when no fallback image has been specified.
+var ErrNoFallbackImage = errors.New("no fallback image has been specified")
+
+// MagicDir is a working directory for envbuilder. We use this to
+// store files that are used when building images.
+type MagicDir string
+
+// String returns the string representation of the MagicDir.
+func (m MagicDir) String() string {
+	if m == "" {
+		// Instead of the zero value, use defaultMagicDir.
+		return defaultMagicDir
+	}
+	return filepath.Join("/", string(m))
+}
+
+// MagicDir implements fmt.Stringer.
+var _ fmt.Stringer = MagicDir("")
+
+// MagicFile is a file that is created in the workspace
+// when envbuilder has already been run. This is used
+// to skip building when a container is restarting.
+// e.g. docker stop -> docker start
+func (m MagicDir) Built() string {
```

Comment:
Bit: I feel like this should return `bool`, and have another method `MarkBuilt()` or something which tries to create the path. I feel like this is currently bleeding implementation details out.

Reply:
While this is true, I think this can be mostly addressed by simply moving it to an internal package.